File: Problems/Graphs/Number-of-Connected-Components-in-an-Undirected-Graph/Number-of-Connected-Components-in-an-Undirected-Graph.cpp
#include <iostream>
#include <vector>
#include <numeric>

// Using union find algorithm - Time: O(V + E)

class unionFind
{
    std::vector<int> id;
    int cnt;

public:
    unionFind(int n) : cnt(n), id(n)
    {
        iota(begin(id), end(id), 0);
    }

    int find(int a)
    {
        return id[a] == a ? a : (id[a] = find(id[a]));
    }

    void connect(int a, int b)
    {
        int p = find(a); 
        int q = find(b);
        
        if (p == q)
            return;
        id[p] = q;

        cnt--;
    }

    int getCount()
    {
        return cnt;
    }
};

class Solution
{
public:
    int countComponents(int n, std::vector<std::vector<int>> &edges)
    {
        unionFind uf(n);
        
        for (auto &e : edges)
        {
            uf.connect(e[0], e[1]);
        }

        return uf.getCount();
    }
};

int main()
{
    int n = 5;
    // std::vector<std::vector<int>> edges = {{0, 1}, {1, 2}, {3,4}};
    std::vector<std::vector<int>> edges = {
        {0, 1},
        {1, 2},
        {2, 3},
        {3, 4}
    };

    std::cout << "Input: n = " << n << ", edges = [";
    for (int i = 0; i < edges.size(); i++)
    {
        std::cout << "[" << edges[i][0] << ", " << edges[i][1] << "]";
        if (i != edges.size() - 1)
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;

    Solution sol;
    int result = sol.countComponents(n, edges);

    std::cout << "Output: " << result << std::endl;

    return 0;
}

File: Problems/Graphs/Number-of-Connected-Components-in-an-Undirected-Graph/Number-of-Connected-Components-in-an-Undirected-Graph.cs
using System;
using System.Linq;
using System.Collections.Generic;

// Using Union find algorithm - Time: O(V + E)

public class UnionFind
{
    private int[] id;
    private int cnt;

    public UnionFind(int n)
    {
        cnt = n;
        id = Enumerable.Range(0, n).ToArray();
    }

    public int Find(int a)
    {
        return id[a] == a ? a : (id[a] = Find(id[a]));
    }

    public void Connect(int a, int b)
    {
        int p = Find(a);
        int q = Find(b);

        if (p == q) 
            return;
        id[p] = q;

        cnt--;
    }

    public int GetCount()
    {
        return cnt;
    }
}

public class Solution
{
    public int CountComponents(int n, int[][] edges)
    {
        UnionFind uf = new UnionFind(n);
        foreach (var e in edges)
        {
            uf.Connect(e[0], e[1]);
        }
        
        return uf.GetCount();
    }
}

class Program
{
    static void Main()
    {
        int n = 5;
        int[][] edges = { 
            new int[] { 0, 1 }, 
            new int[] { 1, 2 }, 
            new int[] { 3, 4 } 
        };

        Console.Write("Input: n = {0}, edges = [", n);
        for (int i = 0; i < edges.Length; i++)
        {
            Console.Write("[{0}, {1}]", edges[i][0], edges[i][1]);
            if (i < edges.Length - 1) 
            {
                Console.Write(",");
            }
        }
        Console.WriteLine("]");

        
        // int[][] edges = { 
        //     new int[] { 0, 1 }, 
        //     new int[] { 1, 2 }, 
        //     new int[] { 2, 3 }, 
        //     new int[] { 3, 4 } 
        // };

        Solution sol = new Solution();
        int result = sol.CountComponents(n, edges);

        Console.WriteLine("Output: " + result);
    }
}
